### PR TITLE
update clusterapi nodegroupset processor

### LINKED ISSUE #5158: Cluster API provider should ignore labels for Alibaba Cloud, IBM Cloud, GKE, and Azure
**Which component are you using?**:

cluster-autoscaler

**Is your feature request designed to solve a problem? If so describe the problem this feature should solve.**:

the experience for balancing similar nodes on the various infrastructures requires the user to know about the labels and invoke the ignore label flag for each.

**Describe the solution you'd like.**:

adding known labels to the cluster-api nodegroupset processor would make this experience much better for users.

**Describe any alternative solutions you've considered.**:

we could create more documentation to expose these labels, and update the helm charts as well.

**Additional context.**:

[cluster-api](https://cluster-api.sigs.k8s.io/) is a project that creates an abstraction around the infrastructure provider resources. as such, it can provide access to many different clouds. this makes the cluster-api provider a little unique in that it will need to ignore more labels in its nodegroupset processor if we want to make a seamless experience for users.

### PR DESCRIPTION
**Pull request by elmiko**
#### Which component this PR applies to?

cluster-autoscaler

#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

this change adds a few more known labels to the processor for
cluster-api, in specific it adds more support for IBM Cloud,
Alibaba Cloud, GCP, and Azure.

#### Which issue(s) this PR fixes:

Fixes #5158

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by enxebre** [comment]
Hey Mike, could you create an issue providing context on the problem the PR solves?

File: cluster-autoscaler/processors/nodegroupset/clusterapi_nodegroups.go [hunk 1]
```diff
@@ -25,11 +25,13 @@ import (
 // even if they have different infrastructure provider-specific labels.
 func CreateClusterAPINodeInfoComparator(extraIgnoredLabels []string) NodeInfoComparator {
 	capiIgnoredLabels := map[string]bool{
-		"topology.ebs.csi.aws.com/zone":                 true, // this is a label used by the AWS EBS CSI driver as a target for Persistent Volume Node Affinity
-		"topology.diskplugin.csi.alibabacloud.com/zone": true, // this is a label used by the Alibaba Cloud CSI driver as a target for Persistent Volume Node Affinity
-		"ibm-cloud.kubernetes.io/worker-id":             true, // this is a label used by the IBM Cloud Cloud Controler Manager
+		"ibm-cloud.kubernetes.io/vpc-instance-id":       true, // used on IBM Cloud when a VPC is in use
```
Comment (enxebre) [inline cluster-autoscaler/processors/nodegroupset/clusterapi_nodegroups.go:28]: can we may be expose them from each providers and reference them here so we don't need to manually keep the list?

**Comment by sbueringer** [comment]
@mdbooth I think there are similar labels for OpenStack (more details in the recording of the CAPI Office hours 31th August)

**Comment by elmiko** [comment]
> Hey Mike, could you create an issue / update the PR desc providing context on the problem the PR solves?

absolutely, i will create one today

File: cluster-autoscaler/processors/nodegroupset/clusterapi_nodegroups.go (hunk 1 above)
Comment (elmiko) [reply cluster-autoscaler/processors/nodegroupset/clusterapi_nodegroups.go:28]: i would be ok with modifying the other nodegroupset processors to expose the variables as constants there, would that work?

Comment (elmiko) [reply cluster-autoscaler/processors/nodegroupset/clusterapi_nodegroups.go:28]: created #5158

**Comment by elmiko** [comment]
created #5158

File: cluster-autoscaler/processors/nodegroupset/clusterapi_nodegroups.go (hunk 1 above)
Comment (elmiko) [reply cluster-autoscaler/processors/nodegroupset/clusterapi_nodegroups.go]: changed to use consts from other providers where applicable

**Comment by elmiko** [comment]
update
* created const values for other providers ignored labels
* change clusterapi implementation to use consts
* added more labels to clusterapi processor

this update changed the structure here a little, but the core is still the same.

**Comment by elmiko** [comment]
updated to fix linting errors

**Comment by fabriziopandini** [comment]
@elmiko now that I better understand this better (thanks for the overview) I think we should lead cluster API users toward a set of recommendations that are driven by the less surprise principle, that IMO is "when using node-groups load balancing, please explicitly set the balancing-labels you want to use" because this avoids node-groups load balancing to be compromised if someone for any reason add a new label to nodes.
I this idea makes sense and gains consensus, then we can eventually stop maintaining this list

**Comment by elmiko** [comment]
+1 @fabriziopandini 

for reference, we discussed this topic at a special meeting today, https://www.youtube.com/watch?v=jbhca_9oPuQ , and it seems like the community would like to go in the direction that @fabriziopandini is mentioning.

i am going to re-evaluate our work on the nodegroupset processor, most likely remove the cluster-api custom bits, and then propose improved documentation instead.

/hold

**Comment by elmiko** [comment]
i am going to close this in favor of the option we discussed previously (removing the processor for capi)

**Pull request closed without merging**